File: src/main/java/com/gugugu/oritech/client/gl/Shader.java
package com.gugugu.oritech.client.gl;

import com.gugugu.oritech.util.Side;
import com.gugugu.oritech.util.SideOnly;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static org.lwjgl.opengl.GL20C.*;

/**
 * @author theflysong
 * @author squid233
 * @since 1.0
 */
@SideOnly(Side.CLIENT)
public class Shader implements AutoCloseable {
    private final int programId;
    private final Map<CharSequence, GLUniform> uniformMap = new HashMap<>();

    public Shader(CharSequence vertexShader, CharSequence fragmentShader) {
        programId = glCreateProgram();
        int vsh = compileShader(GL_VERTEX_SHADER, "vertex", vertexShader);
        int fsh = compileShader(GL_FRAGMENT_SHADER, "fragment", fragmentShader);
        glAttachShader(programId, vsh);
        glAttachShader(programId, fsh);
        glLinkProgram(programId);
        if (glGetProgrami(programId, GL_LINK_STATUS) == GL_FALSE) {
            throw new IllegalStateException("Failed to link program " +
                                            programId +
                                            ". Log: " +
                                            glGetProgramInfoLog(programId));
        }
        glDetachShader(programId, vsh);
        glDetachShader(programId, fsh);
        glDeleteShader(vsh);
        glDeleteShader(fsh);

        // Initialize uniforms
        createUniform("ColorModulator", UniformType.F4).ifPresent(uniform -> uniform.set(1.0f, 1.0f, 1.0f, 1.0f));
    }

    private static int compileShader(int type, String typeStr, CharSequence src) {
        int shader = glCreateShader(type);
        glShaderSource(shader, src);
        glCompileShader(shader);
        if (glGetShaderi(shader, GL_COMPILE_STATUS) == GL_FALSE) {
            throw new IllegalStateException("Failed to compile " +
                                            typeStr +
                                            " shader. Log: " +
                                            glGetShaderInfoLog(shader));
        }
        return shader;
    }

    public Optional<GLUniform> createUniform(CharSequence name,
                                             UniformType type) {
        if (uniformMap.containsKey(name)) {
            return Optional.of(uniformMap.get(name));
        }
        int loc = glGetUniformLocation(programId, name);
        if (loc != -1) {
            GLUniform uniform = new GLUniform(loc, type);
            uniformMap.put(name, uniform);
            return Optional.of(uniform);
        }
        return Optional.empty();
    }

    public Optional<GLUniform> getUniform(CharSequence name) {
        return Optional.ofNullable(uniformMap.get(name));
    }

    public void uploadUniforms() {
        for (GLUniform uniform : uniformMap.values()) {
            uniform.upload();
        }
    }

    public void bind() {
        glUseProgram(programId);
    }

    public void unbind() {
        glUseProgram(0);
    }

    public void delete() {
        glDeleteProgram(programId);
    }

    @Override
    public void close() {
        for (GLUniform uniform : uniformMap.values()) {
            uniform.close();
        }
        delete();
    }

    public int getProgramId() {
        return programId;
    }
}

File: src/Shader.cpp
#include <Shader.h>

#include <fstream>
#include <sstream>
#include <iostream>

#define LOG_SIZE (1024)

Shader::Shader(std::string vertexShader, std::string fragmentShader)
    : programId(glCreateProgram()) {
    std::string vertexShaderCode, fragmentShaderCode;

    std::ifstream vertexShaderFile, fragmentShaderFile;
    std::stringstream vertexShaderStream, fragmentShaderStream;

    vertexShaderFile.exceptions(std::ifstream::failbit | std::ifstream::badbit);
    fragmentShaderFile.exceptions(std::ifstream::failbit | std::ifstream::badbit);

    try {
        vertexShaderFile.open(vertexShader.c_str());
        fragmentShaderFile.open(fragmentShader.c_str());

        vertexShaderStream << vertexShaderFile.rdbuf();
        fragmentShaderStream << fragmentShaderFile.rdbuf();

        vertexShaderFile.close();
        fragmentShaderFile.close();

        vertexShaderCode = vertexShaderStream.str();
        fragmentShaderCode = fragmentShaderStream.str();
    }
    catch (std::ifstream::failure e) {
        std::cout << "Couldn't read shader file!" << std::endl;
        return;
    }
    const char *vertexShaderBuffer = vertexShaderCode.c_str(), *fragmentShaderBuffer = fragmentShaderCode.c_str();

    //------------------------------------------

    int success;
    char log[LOG_SIZE];

    GLuint vertexShaderId = glCreateShader(GL_VERTEX_SHADER), fragmentShaderId = glCreateShader(GL_FRAGMENT_SHADER);
    glShaderSource(vertexShaderId, 1, &vertexShaderBuffer, NULL);
    glCompileShader(vertexShaderId);

    glGetShaderiv(vertexShaderId, GL_COMPILE_STATUS, &success);
    if(success == 0) {
        glGetShaderInfoLog(vertexShaderId, LOG_SIZE, NULL, log);
        std::cout << "Vertex Shader Compilation failed!\n" << log << std::endl;
        return;
    }

    glShaderSource(fragmentShaderId, 1, &fragmentShaderBuffer, NULL);
    glCompileShader(fragmentShaderId);

    glGetShaderiv(fragmentShaderId, GL_COMPILE_STATUS, &success);
    if(success == 0) {
        glGetShaderInfoLog(fragmentShaderId, LOG_SIZE, NULL, log);
        std::cout << "Fragment Shader Compilation failed!\n"  << log << std::endl;
        return;
    }

    //------------------------------------------

    glAttachShader(this->programId, vertexShaderId);
    glAttachShader(this->programId, fragmentShaderId);
    glLinkProgram(this->programId);

    glGetProgramiv(this->programId, GL_LINK_STATUS, &success);
    if (success == 0) {
        glGetProgramInfoLog(this->programId, LOG_SIZE, NULL, log);
        std::cout << "Shader Program Linking failed!\n"  << log << std::endl;
        return;
    }

    glDeleteShader(vertexShaderId);
    glDeleteShader(fragmentShaderId);
}

Shader::~Shader() {
    glDeleteProgram(this->programId);
}

void Shader::Use() {
    glUseProgram(this->programId);
}